sql-to-dbsp-compiler/lib/sqllib/src/lib.rs

```diff
@@ -18,6 +19,291 @@ use std::ops::Add;
+            }
+        }
+    }
+}
+
+// Macro to create variants of a function with 3 arguments
+// If there exists a function is f___(x: T, y: S, z: V) -> U, this creates
+// seven functions:
+// - f__N(x: T, y: S, z: Option<V>) -> Option<U>
+// - f_N_(x: T, y: Option<S>, z: V) -> Option<U>
+// - etc.
+// The resulting functions return Some only if all arguments are 'Some'.
+#[macro_export]
+macro_rules! some_function3 {
+    ($func_name:ident, $arg_type0:ty, $arg_type1:ty, $arg_type2: ty, $ret_type:ty) => {
+        ::paste::paste! {
+            pub fn [<$func_name __N>]( arg0: $arg_type0, arg1: $arg_type1, arg2: Option<$arg_type2> ) -> Option<$ret_type> {
+                let arg2 = arg2?;
+                Some([<$func_name ___>](arg0, arg1, arg2))
+            }
+
+            pub fn [<$func_name _N_>]( arg0: $arg_type0, arg1: Option<$arg_type1>, arg2: $arg_type2 ) -> Option<$ret_type> {
+                let arg1 = arg1?;
+                Some([<$func_name ___>](arg0, arg1, arg2))
+            }
+
+            pub fn [<$func_name _NN>]( arg0: $arg_type0, arg1: Option<$arg_type1>, arg2: Option<$arg_type2> ) -> Option<$ret_type> {
+                let arg1 = arg1?;
+                let arg2 = arg2?;
+                Some([<$func_name ___>](arg0, arg1, arg2))
+            }
+
+            pub fn [<$func_name N__>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: $arg_type2 ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                Some([<$func_name ___>](arg0, arg1, arg2))
+            }
+
+            pub fn [<$func_name N_N>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: Option<$arg_type2> ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                let arg2 = arg2?;
+                Some([<$func_name ___>](arg0, arg1, arg2))
+            }
+
+            pub fn [<$func_name NN_>]( arg0: Option<$arg_type0>, arg1: Option<$arg_type1>, arg2: $arg_type2 ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                let arg1 = arg1?;
+                Some([<$func_name ___>](arg0, arg1, arg2))
+            }
+
+            pub fn [<$func_name NNN>]( arg0: Option<$arg_type0>, arg1: Option<$arg_type1>, arg2: Option<$arg_type2> ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                let arg1 = arg1?;
+                let arg2 = arg2?;
+                Some([<$func_name ___>](arg0, arg1, arg2))
+            }
+        }
+    }
+}
+
+// Macro to create variants of a function with 2 arguments
+// If there exists a function is f_t_t(x: T, y: T) -> U, this creates
+// three functions:
+// - f_tN_t(x: T, y: Option<T>) -> Option<U>
+// - f_t_tN(x: Option<T>, y: T) -> Option<U>
+// - f_tN_tN(x: Option<T>, y: Option<T>) -> Option<U>
+// The resulting functions return Some only if all arguments are 'Some'.
+#[macro_export]
+macro_rules! some_operator {
+    ($func_name: ident, $short_name: ident, $arg_type: ty, $ret_type: ty) => {
+        ::paste::paste! {
```

Comment:
I see that these are somehow for operators, but what's the difference between this and some_function2?